### PR TITLE
improve progress display when running apps; speed up startup

### LINKED ISSUE #9435: multi-second delay before first progress message from 'flutter run'
On Linux at 1df6917b60b38115a436d6bddaddacb087ea0fb4

To reproduce:
```
flutter create click
cd click
flutter run
```

I see a multi-second delay before the first progress message:

```
Launching lib/main.dart on Android SDK built for x86 64 in debug mode...
```

This is rather disconcerting, particularly when launching from within IDEA where UI responsiveness is expected. We should print the first progress message sooner, so the user knows to be patient.

### PR DESCRIPTION
**Pull request by devoncarew**
- fix https://github.com/flutter/flutter/issues/9435; speed-up app startup and move our initial message to before we do a lot of work
- cache the result of calculating GradleAppOutDirV2; this call takes about a second, and is called 4-5 times
- move the `Launching...` message before we do a lot of android/gradle work
- change the `installApp()` call to be async, so we can use async methods within, so the spinner for the `Installing build/app/outputs/apk/app.apk... ` message doesn't appear frozen

Start-up text after the PR:

```
Launching lib/main.dart on Nexus 5X in debug mode...
Running 'gradle assembleDebug'...                     7.9s
Built build/app/outputs/apk/app-debug.apk (27.9MB).
Installing build/app/outputs/apk/app.apk...           6.9s
Syncing files to device...                            3.4s

🔥  To hot reload your app on the fly, press "r". To restart the app entirely, press "R".
The Observatory debugger and profiler is available at: http://127.0.0.1:8104/.
For a more detailed help message, press "h". To quit, press "q".
```

### CODE REVIEW

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 1]
```diff
@@ -233,16 +233,17 @@ class AndroidDevice extends Device {
...
-    final String failure = failureExp.stringMatch(installOut);
+    final String failure = failureExp.stringMatch(installResult.stdout);
     if (failure != null) {
       printError('Package install error: $failure');
       return false;
     }
 
-    runCheckedSync(adbCommandForDevice(<String>['shell', 'echo', '-n', _getSourceSha1(app), '>', _getDeviceSha1Path(app)]));
+    await runCheckedAsync(adbCommandForDevice(
+      <String>['shell', 'echo', '-n', _getSourceSha1(app), '>', _getDeviceSha1Path(app)]));
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/android/android_device.dart]: nit: newline before `));` to match newline after matching `(`.

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 2]
```diff
@@ -63,12 +65,19 @@ String get gradleAppOut {
...
-      return '$gradleAppOutDirV2/app.apk';
+      return '${getGradleAppOutDirV2()}/app.apk';
   }
   return null;
 }
 
-String get gradleAppOutDirV2 {
+String getGradleAppOutDirV2() {
+  if (_cachedGradleAppOutDirV2 == null)
+    _cachedGradleAppOutDirV2 = _calculateGradleAppOutDirV2();
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/android/gradle.dart]: nit: use `??=`

File: packages/flutter_tools/lib/src/run_cold.dart [hunk 3]
```diff
@@ -158,7 +158,7 @@ class ColdRunner extends ResidentRunner {
   void printHelp({ @required bool details }) {
     bool haveDetails = false;
     if (_result.hasObservatory)
-      printStatus('The Observatory debugger and profiler is available at: ${_result.observatoryUri}');
+      printStatus('The Observatory debugger and profiler is available at: ${_result.observatoryUri}.');
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/run_cold.dart]: the lack of period here is intentional, we don't want the url to get a trailing period.
The grammar is fine, the `:` ends the sentence.

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 4]
```diff
@@ -578,7 +578,7 @@ class HotRunner extends ResidentRunner {
       ansiAlternative: '$red$fire$bold  To hot reload your app on the fly, '
                        'press "r". To restart the app entirely, press "R".$reset'
     );
-    printStatus('The Observatory debugger and profiler is available at: $_observatoryUri');
+    printStatus('The Observatory debugger and profiler is available at: $_observatoryUri.');
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/run_hot.dart]: as above

**Comment by Hixie** [comment]
LGTM. Bonus points for tests, if anything is testable here.

File: packages/flutter_tools/lib/src/android/android_device.dart (hunk 1 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/android/android_device.dart]: done

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 2 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/android/gradle.dart]: done

File: packages/flutter_tools/lib/src/run_cold.dart (hunk 3 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/run_cold.dart]: ok, removing here and below

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 4 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/run_hot.dart]: done

**Comment by devoncarew** [comment]
> LGTM. Bonus points for tests, if anything is testable here.

Thanks! In terms of tests, I think the best thing would be to have a benchmark for app startup time. cc'ing @yjbanov to see if that sounds reasonable.

**Review by jakobr-google — APPROVED** [review]
LGTM

**Comment by johnmccutchan** [comment]
LGTM. But, I'm not sure why all the `runCheckedSync` calls were converted into `await runCheckedAsync`. Asynchronous code is slower and uses more memory. If you're going to immediately await the result, your program is still effectively synchronous around that point.

**Review by johnmccutchan — APPROVED** [review]

**Comment by devoncarew** [comment]
The progress display on the command-line uses timers every 100ms to tick the UI forward. Long synchronous calls will mean we don't get our timer calls, and the progress spinner appears frozen.

**Comment by yjbanov** [comment]
> In terms of tests, I think the best thing would be to have a benchmark for app startup time. 

We probably maxed out on the number of columns in our build table on the dashboard, but we could definitely expand the `hot_mode_dev_cycle__benchmark` test to submit more metrics.

**Pull request merged**